app/scenes/Document/Shared.tsx

```diff
@@ -16,38 +16,77 @@ import Loading from "./components/Loading";
 
 const EMPTY_OBJECT = {};
 
+type Response = {
+  document: DocumentModel;
+  sharedTree?: NavigationNode | undefined;
+};
+
 type Props = RouteComponentProps<{
   shareId: string;
   documentSlug: string;
 }> & {
   location: Location<{ title?: string }>;
 };
 
+/**
+ * Find the document UUID from the slug given the sharedTree
+ *
+ * @param documentSlug The slug from the url
+ * @param response The response payload from the server
+ * @returns The document UUID, if found.
+ */
+function useDocumentId(documentSlug: string, response?: Response) {
+  let documentId;
+
+  function findInTree(node: NavigationNode) {
+    if (node.url.endsWith(documentSlug)) {
+      documentId = node.id;
+      return true;
+    }
+    if (node.children) {
+      for (const child of node.children) {
+        if (findInTree(child)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  if (response?.sharedTree) {
+    findInTree(response.sharedTree);
+  }
+
+  return documentId;
+}
+
 function SharedDocumentScene(props: Props) {
   const { ui } = useStores();
-
   const theme = useTheme();
-  const [response, setResponse] = React.useState<{
-    document: DocumentModel;
-    sharedTree?: NavigationNode | undefined;
-  }>();
+  const [response, setResponse] = React.useState<Response>();
   const [error, setError] = React.useState<Error | null | undefined>();
   const { documents } = useStores();
   const { shareId, documentSlug } = props.match.params;
+  const documentId = useDocumentId(documentSlug, response);
 
   // ensure the wider page color always matches the theme
   React.useEffect(() => {
     window.document.body.style.background = theme.background;
   }, [theme]);
 
+  React.useEffect(() => {
+    if (documentId) {
+      ui.setActiveDocument(documentId);
+    }
+  }, [ui, documentId]);
+
   React.useEffect(() => {
     async function fetchData() {
       try {
         const response = await documents.fetchWithSharedTree(documentSlug, {
           shareId,
         });
         setResponse(response);
-        ui.setActiveDocument(response.document);
```

Comment:
I made this change — I think it's correct to have only one call to setActiveDocument. Otherwise there's some strange looking behavior when there's a bit of latency and two successive navigation clicks.